gateway/src/apicast/proxy.lua

```diff
@@ -263,16 +289,22 @@ function _M:rewrite(service)
   local ctx = ngx.ctx
   local var = ngx.var
 
+  local parsed_usage = parse_usage(usage)
+
   -- save those tables in context so they can be used in the backend client
-  ctx.usage = usage_params
+  ctx.usage = parsed_usage
```

Comment:
This is something we will change in the future, right? Exposing the real `usage` object rather than this. 

How the SOAP policy could add usage when the `Usage` object is not exposed?

Reply:
Yes. This is related to what I mentioned in the PR description.

There are some changes we need to make here. First, we need to Proxy needs to receive the context via params in `.rewrite`. Second, it will need to merge with the usage in the context rather than just assigning it, and third, we'll need to move this formatting closer to the call to the backend client call. The context will always contain instances of Usage.

I realized that making the changes I mentioned above some tests break because they're too tied to the implementation details.

As you can see, in order to properly address this I'll need to change a few things and I felt there was already too much in this PR. I left it for the next so we can focus here in the MappingRulesMatcher and Usage that will be exposed and potentially used from other modules.

Reply:
I forgot to mention that the current version of the code also saves in `ctx.usage` the formatted version.

Reply:
👍

Reply:
I added 3 new commits that fix what we discussed. In the end I decided to include them here to make sure that the approach taken works well.